test/abstractarray.jl

```diff
@@ -797,3 +797,34 @@ end
 
 # dispatch loop introduced in #19305
 @test [(1:2) zeros(2,2); ones(3,3)] == [[1,2] zeros(2,2); ones(3,3)] == [reshape([1,2],2,1) zeros(2,2); ones(3,3)]
+
+# Test a very basic linear interpolations-like custom array
+immutable InterpolatedVector{T,A} <: AbstractVector{T}
+    data::A
+end
+InterpolatedVector{T}(data::AbstractVector{T}) = InterpolatedVector{T,typeof(data)}(data)
+Base.size(V::InterpolatedVector) = size(V.data)
+# This custom type supports all Numbers as scalar indices; protect them in a 0-d array for non-scalar methods
+Base.to_index(::InterpolatedVector, x::Number) = fill(x) # could be a specialized immutable
+Base.indexed_eltype{T}(V::InterpolatedVector{T}, xs...) = _reduce_promote_eltype(T, xs...)
```

Comment:
You could use `Base.promote_eltype(T, xs...)` instead of `_reduce_promote_eltype`. Unfortunately, that doesn't help with the slowness of inference over the `view`s you're seeing.